added microk8s installation docs (#285)
codacy/chart: docs/infrastructure/microk8s-quickstart.md

`helm` is the client facing side, while `tiller` is the server/cluster facing side.

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster. For your infrastructure, you will need the following:

-   A machine running Ubuntu 18.04 LTS. You must start a local or remote command line session on this machine.
-   A [PostgreSQL instance with all the necessary databases created](../requirements.md#postgresql-server-setup). The machine above must be able to connect to this PostgreSQL instance.

All the following steps assume that you are starting from a blank slate.

## 3. Installing microk8s

1.  Make sure the machine has the `nfs-common` package installed.

    ```bash
    sudo apt update && sudo apt install nfs-common -y
    ```

2.  Install microk8s from the `1.15/stable` channel.

    ```bash
    sudo snap install microk8s --classic --channel=1.15/stable && \
    sudo usermod -a -G microk8s $USER  && \
    sudo su - $USER
    ```

    Check that microk8s is running.

    ```bash
    microk8s.status --wait-ready
    ```

> 3.  Install the version `v2.16.3` helm binary

Reviewer: ```suggestion
3.  Install the version `v2.16.3` of the helm binary
```

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster.
For your infrastructure, you will need the following:

1.  A machine running Ubuntu 18.04 LTS.
2.  A postgres database with all the necessary databases created and open firewall rule. ([Documentation can be found here.](../requirements.md#postgresql-server-setup))
3.  Establish a session (local or remote) onto the machine described in 1. above.

All the following steps assume that you are starting from a blank slate.

## 3. Installing microk8s

1.  Make sure the machine is up to date and also has the `nfs-common` package installed

    ```bash
    sudo apt update && sudo apt install nfs-common -y
    ```

2.  Install the `1.15/stable` channel of microk8s.

    ```bash
    sudo snap install microk8s --classic --channel=1.15/stable && \
    sudo usermod -a -G microk8s $USER  && \
    su - $USER
    ```

    Check if everything was installed correctly.

    ```bash
    microk8s.status --wait-ready
    ```

3.  Install the helm binary

    ```bash
    HELM_PKG=helm-v2.16.3-linux-amd64.tar.gz
    wget https://get.helm.sh/$HELM_PKG
    tar xvzf $HELM_PKG

>     mv linux-amd64/tiller linux-amd64/helm /usr/local/bin

Reviewer 2: confirm if this needs sudo.. according to andrea it does

Reviewer 3: It does, so that you can write on the folder /usr/local/bin.

Two executables will get installed onto the cluster as part of this process: `helm` and `tiller`.

* `helm` is responsible for resolving the configuration of the chart to be installed, while issuing the correct install commands onto the cluster.
* `tiller` is responsible for receiving the install commands issued by `helm`, as well as managing the lifecycle of the components that have been installed.

`helm` is the client facing side, while `tiller` is the server/cluster facing side.

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster.
For your infrastructure, you will need the following:

1. A machine running Ubuntu 18.04 LTS.
2. A postgres database and open firewall rule. ([Documentation can be found here.](../requirements.md))

To start installing Codacy, you will need to:

1. Establish a session (local or remote) onto the machine described in 1. above.

Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, consider that Codacy is supported up to kubernetes 1.15. Please jump on to the section [Configuring microk8s](###4.-Configuring-microk8s) if you are on a compatible kubernetes version.

> ## 3. Installing microk8s

Reviewer 2: suggestion from rodrigo:
Single script with verification at the begging if we are on the right OS, kernel, etc

1. Make sure the machine is up to date and also has the `nfs-common` package installed

   ```bash
   sudo apt update && apt install nfs-common -y

> 2. Install microk8s. Codacy is supported up to kubernetes 1.15. Therefore we recommend following the [official documentation on how to install microk8s](https://microk8s.io/docs/) with the `1.15/stable` channel.

Reviewer 2: We need to freeze also the version of the microk8s itself

> 2. Install microk8s. Codacy is supported up to kubernetes 1.15. Therefore we recommend following the [official documentation on how to install microk8s](https://microk8s.io/docs/) with the `1.15/stable` channel.

Reviewer 4: Please mention this version before the link to the microk8s repo at the top. I followed the quickstart they have there and had to redo everything

Reviewer 4: Maybe we should add the command needed to do it?

Reviewer 2: Done

3. Allow privileged containers for microk8s

   ```bash

>    sudo echo "--allow-privileged=true" >> /var/snap/microk8s/current/args/kube-apiserver

Reviewer 4: This step gave me permission denied, even with sudo. Might've been something to do with the instance I'm using, but maybe we should take a look.

Reviewer 3: This has to do with the fact that the (global) directories under /var/snap are actually read-only mount points (see https://askubuntu.com/a/1035241).

So, we should have another way of doing this configuration, otherwise the workaround will be rather complex. Just an idea, maybe the snap could be installed in the user space instead, which would allow changing the script (and could we enforce the customers to do this)?...

Reviewer 2: The command on installing microk8s should solve this `sudo usermod -a -G microk8s $USER `

   ```

## 4. Configuring microk8s

> 1. First, we must enable the following plugins on microk8s:

Reviewer 4: These steps take a bit of time, maybe it's worth it to mention it somewhere?

    ```bash
   microk8s.enable dns &&
   microk8s.enable storage &&
   microk8s.enable ingress &&
   microk8s.status --wait-ready
    ```

2. Now that these plugins have been enabled, we should restart microk8s to ensure a proper bootstrap of the cluster

    ```bash
   microk8s.stop && microk8s.start && microk8s.status --wait-ready
    ```

    Further information on this subject can be [found here](https://github.com/ubuntu/microk8s/issues/493#issuecomment-498167435).

3. The plugins are now enabled and the cluster bootstrapped. However, we must still wait for some microk8s internals (dns, http, and ingress) plugins to be ready. Failing to do so can result in pods entering a `CrashLoopBackoff` state:

   ```bash

>     microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=kube-dns

Reviewer 4: From microk8s documentation, they actually suggest to use an alias for kubectl `alias kubectl='microk8s.kubectl'`.
Do you think we should mention it here?

Reviewer 3: My opinion: it would add an extra step / instruction and the benefits of calling a shorter command could be outweighed by the issue of collision for customers who happen to also use the "standard" kubectl.

Reviewer 3: On the other hand, it might simplify the process of following [these instructions](https://github.com/codacy/chart/blob/master/docs/install.md) with microk8s. We'll consider this idea.

Reviewer 2: Done

    microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=hostpath-provisioner

>     microk8s.kubectl wait -n default --for=condition=Ready pod -l name=nginx-ingress-microk8s

Reviewer 4: The controller is created in the ingress namespace, update this
```suggestion
    microk8s.kubectl wait -n ingress --for=condition=Ready pod -l name=nginx-ingress-microk8s
```

    ```

After these commands return successfully, we have ensured that dns, http, and nginx ingress are enabled and working properly inside the cluster.

## 5. Installing Codacy

### 1. Helm and Tiller

First we must install `helm` onto the cluster. With `helm`, we can easily install charts and manage the lifecycle of the installed artifacts. One example is the ability to rollback a failed install.

> 1. __Codacy supports up to `helm` version 2.16.3.__

Reviewer 4: Should we add the command needed to install with this specific version?

Reviewer 3: Installing Helm as part of preparing the environment to install Codacy is already mentioned here:

https://github.com/codacy/chart/blob/master/docs/index.md#preparing-to-install-codacy

We need to integrate the quickstart for the microk8s better with the main installation process in that page, though.

Reviewer 2: done

   __We currently do not support `helm` v3.__
2. Follow the instructions on installing the `helm` client on the [Preparing to install Codacy documentation](../../docs/index.md).

3. We must now deploy `tiller` onto the cluster.
   1. First we must set a `serviceaccount` and a `clusterrolebinding` for `tiller` on the microk8s cluster:

      ```bash
      microk8s.kubectl create serviceaccount --namespace kube-system tiller
      microk8s.kubectl create clusterrolebinding tiller-cluster-rule --clusterrole=cluster-admin --serviceaccount=kube-system:tiller
      ```

   2. Then we can install `tiller`, which will be installed with the same `HELM_VERSION` that you have specified earlier and we wait for tiller to be ready:

      ```bash

>       helm init --service-account tiller

Reviewer 4: I was having some troubles with this and had to do
`mkdir ~/.kube && sudo chown -Rv ubuntu:ubuntu ~/.kube && sudo microk8s.config > .kube/config`
before
`helm init --service-account tiller`
to make it work

      microk8s.kubectl -n kube-system wait --for=condition=Ready pod -l name=tiller
      ```

### 2. Codacy chart

Please refer to [our chart installation documentation](https://github.com/codacy/chart/blob/master/docs/install.md).

> Additionally, any `kubectl` command from [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md) must be executed as a `microk8s.kubectl` command.

Reviewer 4: As I said above, from microk8s documentation, they actually suggest to use an alias for kubectl commands, consider adding that here

#### 1. Default Values

> We provide a base [values file](##values-microk8s.yaml) that you can use for your microk8s Codacy installation. This file provisions a smaller number of component replicas with limited resources (i.e. CPU and memory). Should you want to use this as a template, please pass your `values-microk8s.yaml` file on (4.) of [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md). Lastly, it is important to correctly set your `codacy-api.ingress.hosts.host` if you are using this file as a template.

Reviewer 4: This is a bit weird, we refer to the chart installation documentation and then we say we should use `values-microk8s.yaml`, we should mention that we need to use this one instead of `values-production.yaml`

#### 2. External Databases

> If you do not have an external database set up for you by an Ops team,  you should follow the [steps described here](../requirements.md).

Reviewer 4: There's an extra space here
```suggestion
If you do not have an external database set up for you by an Ops team, you should follow the [steps described here](../requirements.md).
```

* The address of the cluster (e.g. https://104.248.34.242:164309).
* The user friendly name of the cluster (e.g. microk8s-cluster).
* A reference to the current context. This is the active cluster against which any kubernetes control (`kubectl`) command you run will get executed.

### Helm and tiller

Two executables will get installed onto the cluster as part of this process: `helm` and `tiller`.

* `helm` is responsible for resolving the configuration of the chart to be installed, while issuing the correct install commands onto the cluster.
* `tiller` is responsible for receiving the install commands issued by `helm`, as well as managing the lifecycle of the components that have been installed.

`helm` is the client facing side, while `tiller` is the server/cluster facing side.

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster:

1. A machine running Ubuntu 18.04 LTS.

> 2. Establish a session (local or remote) onto this machine.

Reviewer 2: An RDS and firewall open (maybe a direct link from here to the postgres requirements)

Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, please jump on to the section [Configuring microk8s](###4.-Configuring-microk8s).

## 3. Installing microk8s

1. Make sure the machine is up to date and also has the `nfs-common` package installed

   ```bash
   sudo apt update && apt install nfs-common -y

2. Install microk8s. Codacy is supported up to kubernetes 1.15. Therefore we recommend following the [official documentation on how to install microk8s](https://microk8s.io/docs/) with the `1.15/stable` channel.

## 4. Configuring microk8s

1. First, we must enable the following plugins on microk8s:

    ```bash

>     microk8s.enable dns && microk8s.status --wait-ready

Reviewer: the `microk8s.status --wait-ready` is redundant since it's performed at the end

      ```bash
      helm init --service-account tiller
      microk8s.kubectl -n kube-system wait --for=condition=Ready pod -l name=tiller
      ```

### 2. Codacy chart

Please refer to [our chart installation documentation](https://github.com/codacy/chart/blob/master/docs/install.md).

Additionally, any `kubectl` command from [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md) must be executed as a `microk8s.kubectl` command.

#### 1. Default Values

We provide a base (values file)[##values-microk8s.yaml] that you can use for your microk8s Codacy installation. This file provisions a smaller number of component replicas with limited resources (i.e. CPU and memory). Should you want to use this as a template, please pass your `values-microk8s.yaml` file on (4.) of [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md). Lastly, it is important to correctly set your `codacy-api.ingress.hosts.host` if you are using this file as a template.

#### 2. External Databases

> If you have decided to go with the [default values for the installation](####1.-Default-Values), your microk8s cluster will contain a PostgreSQL database within itself. This is not recommended as the data stored in the database will not be as durable if you store it in an external database.

Reviewer 5: We really should not make it easy for people to install using the internal postgres, because it will end up causing them lots of problems

Author: I agree. I am removing this paragraph.

3. Traffic forwarding must be configured on the machine so that the local microk8s cluster pods can reach the internet, again because of using an AWS EC2 instance. This can be done by running `iptables -P FORWARD ACCEPT` (more info on this subject can be [found here](https://microk8s.io/docs/troubleshooting#common-issues).

## 4. Configuring microk8s

1. First, we must enable the following plugins on microk8s:
   1. `sudo microk8s.enable dns && sudo microk8s.status --wait-ready`
   2. `sudo microk8s.enable storage && sudo microk8s.status --wait-ready`
   3. `sudo microk8s.enable ingress &&  sudo microk8s.status --wait-ready`
2. Now that these plugins have been enabled, we should restart microk8s to ensure a proper bootstrap of the cluster: `microk8s.stop && microk8s.start && microk8s.status --wait-ready`. Further information on this subject can be [found here](https://github.com/ubuntu/microk8s/issues/493#issuecomment-498167435).
3. The plugins are now enabled and the cluster bootstrapped. However, we must still wait for some microk8s internals (dns, http, and ingress) plugins to be ready. Failing to do so can result in pods entering a `CrashLoopBackoff` state:

>    1. ```microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=kube-dns --timeout=300s && microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=hostpath-provisioner --timeout=300s && microk8s.kubectl wait -n default --for=condition=Ready pod -l app=default-http-backend --timeout=300s && microk8s.kubectl wait -n default --for=condition=Ready pod -l name=nginx-ingress-microk8s --timeout=300s```

Reviewer 2: ```suggestion
    
   ```microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=kube-dns --timeout=300s && \
   microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=hostpath-provisioner --timeout=300s && \
   microk8s.kubectl wait -n default --for=condition=Ready pod -l app=default-http-backend --timeout=300s && \
   microk8s.kubectl wait -n default --for=condition=Ready pod -l name=nginx-ingress-microk8s --timeout=300s```
```

After runing these commands, we have ensured that dns, http, and nginx ingress are enabled and working properly inside the cluster.

## 5. Installing Codacy

### 1. Helm and Tiller

First we must install `helm` onto the cluster. With `helm`, we can easily install charts and manage the lifecycle of the installed artifacts. One example is the ability to rollback a failed install.
1. Assuming you would want to install `helm` version 2.16.3, export the following environment variable: `export HELM_VERSION=2.16.3`
   1. Please note that we currently do *not* support helm v3.
2. Run this script that will retrieve the desired `helm` version and install it onto the machine's `/usr/local/bin`: `HELM_PKG=helm-${HELM_VERSION}-linux-amd64.tar.gz
wget https://get.helm.sh/$HELM_PKG
tar xvzf $HELM_PKG
su mv linux-amd64/tiller linux-amd64/helm /usr/local/bin

> rm -rvf $HELM_PKG linux-amd64/`.

Reviewer 2: ```suggestion
2. Run this script that will retrieve the desired `helm` version and install it onto the machine's `/usr/local/bin`: 

    ```bash
    $ HELM_VERSION=2.16.3 
    $ HELM_PKG=helm-${HELM_VERSION}-linux-amd64.tar.gz
    $ wget https://get.helm.sh/$HELM_PKG
    $ tar xvzf $HELM_PKG
    $ su mv linux-amd64/tiller linux-amd64/helm /usr/local/bin
    $ rm -rvf $HELM_PKG linux-amd64/
    ```

```

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster:

1. a machine running Ubuntu 18.04 LTS version.
2. establish a remote session (ssh) onto this machine.

Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, please jump on to the [Configuring microk8s](###Configuring-microk8s) section.

## 3. Installing microk8s

You can follow the official documentation on how to [install microk8s](https://microk8s.io/docs/). However, because we were running off of an AWS EC2 instance, we had do the following steps becore installing microk8s:

1. Make sure the machine is up to date and also has the `nfs-common` package installed: `apt update && apt install nfs-common -y`.
2. We must run configure the instance's list of hosts with localhost. This was because our machine was in fact an AWS EC2 instance, and therefore we must allow the correct resolution of the local microk8s cluster on the kubeconfig. We did so by running `sed -i -e "1s/127.0.0.1 localhost.*/127.0.0.1 localhost $(hostname)/" /etc/hosts`.

> 3. Traffic forwarding must be configured on the machine so that the local microk8s cluster pods can reach the internet, again because of using an AWS EC2 instance. This can be done by running `iptables -P FORWARD ACCEPT` (more info on this subject can be [found here](https://microk8s.io/docs/troubleshooting#common-issues).

Reviewer 6: These commands require root/sudo

1. Assuming you would want to install `helm` version 2.16.3, export the following environment variable: `export HELM_VERSION=2.16.3`
   1. Please note that we currently do *not* support helm v3.
2. Run this script that will retrieve the desired `helm` version and install it onto the machine's `/usr/local/bin`: `HELM_PKG=helm-${HELM_VERSION}-linux-amd64.tar.gz
wget https://get.helm.sh/$HELM_PKG
tar xvzf $HELM_PKG
su mv linux-amd64/tiller linux-amd64/helm /usr/local/bin
rm -rvf $HELM_PKG linux-amd64/`.
4. We must now deploy `tiller` onto the cluster.
   1. First we must set a `serviceaccount` and a `clusterrolebinding` for `tiller` on the microk8s cluster: `microk8s.kubectl create serviceaccount --namespace kube-system tiller
microk8s.kubectl create clusterrolebinding tiller-cluster-rule --clusterrole=cluster-admin --serviceaccount=kube-system:tiller`
   2. Then we can install `tiller`, which will be installed with the same `HELM_VERSION` that you have specified earlier: `helm init --service-account tiller`
   3. We wait for tiller to be ready: `microk8s.kubectl -n kube-system wait --for=condition=Ready pod -l name=tiller --timeout=300s`

> ### 2. Codacy chart

Reviewer 2: We should make use of the other docs as much as possible and put only the diffs here: https://github.com/codacy/chart/blob/master/docs/install.md

Reviewer 2: this sections needs to be as little as possible

1. Create a namespace inside the cluster onto which codacy will be deployed. We use the `codacy` namespace on the commands below, but you are free to use your own value for this namespace. `microk8s.kubectl create namespace codacy`
2. The microk8s cluster must know some docker credentials so that the container docker images can be pulled onto the pods. These credentials must belong to the namespace where codacy is installed. You can do so by running: `microk8s.kubectl create secret docker-registry docker-credentials --docker-username=${DOCKER_USERNAME} --docker-password=${DOCKER_PASSWORD} --namespace codacy`. These credentials will be encoded inside the cluster.
3. For codacy to work, we must set a file that which contains information like how many replicas of a given pod will codacy have, what are the cpu and memory limits for these pods, which secret tokens to use for the filestore, among others.

>    1. Define the following variables

Reviewer 2: This is very high maintenance and should already be covered in the installation docs: https://github.com/codacy/chart/blob/master/docs/install.md

* The address of the cluster (e.g. https://104.248.34.242:164309).
* The user friendly name of the cluster (e.g. microk8s-cluster).
* A reference to the current context. This is the active cluster against which any kubernetes control (`kubectl`) command you run will get executed.

### Helm and tiller

Two executables will get installed onto the cluster as part of this process: `helm` and `tiller`.

* `helm` is responsible for resolving the configuration of the chart to be installed, while issuing the correct install commands onto the cluster.
* `tiller` is responsible for receiving the install commands issued by `helm`, as well as managing the lifecycle of the components that have been installed.

`helm` is the client facing side, while `tiller` is the server/cluster facing side.

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster:

1. A machine running Ubuntu 18.04 LTS.

> 2. Establish a remote SSH session onto this machine.

Reviewer 3: Yes, I agree!

> 2. Establish a remote SSH session onto this machine.

Author: 👍

> Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, please jump on to the section [Configuring microk8s](###4.-Configuring-microk8s).

Author: 👍

> Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, please jump on to the section [Configuring microk8s](###4.-Configuring-microk8s).

Reviewer: super important, we do not support k8s > 16, so you should install microk8s  at a very specific version. cc. @h314to for the help

Author: References in Installing 3.2

## 3. Installing microk8s

1. Make sure the machine is up to date and also has the `nfs-common` package installed

   ```bash

>    sudo apt update && apt install nfs-common -y

Reviewer 3: Must close the fenced code block:

```suggestion
   ```bash
   sudo apt update && apt install nfs-common -y
   ```
```

> 2. Install microk8s

Reviewer 3: I like this idea. As it is now, it seems that the user must do something "Codacy-specific" to install microk8s and then also follow the microk8s installation documentation.

> 2. Install microk8s

Author: The thing is they need to install up to 1.15 microk8s, which translates to kubernetes 1.15. Any version above that (like the latest 1.17) will fail installing codacy.

   >sudo snap install microk8s --classic --channel=1.15/stable
   1. We recommend following the official documentation on how to [install microk8s](https://microk8s.io/docs/).

## 4. Configuring microk8s

1. First, we must enable the following plugins on microk8s:
    >sudo microk8s.enable dns && sudo microk8s.status --wait-ready

    >sudo microk8s.enable storage && sudo microk8s.status --wait-ready

    >sudo microk8s.enable ingress &&  sudo microk8s.status --wait-ready

2. Now that these plugins have been enabled, we should restart microk8s to ensure a proper bootstrap of the cluster
    >microk8s.stop && microk8s.start && microk8s.status --wait-ready

    Further information on this subject can be [found here](https://github.com/ubuntu/microk8s/issues/493#issuecomment-498167435).
3. The plugins are now enabled and the cluster bootstrapped. However, we must still wait for some microk8s internals (dns, http, and ingress) plugins to be ready. Failing to do so can result in pods entering a `CrashLoopBackoff` state:

>     >microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=kube-dns --timeout=300s

Author: I don't think so because one is waiting for kube-system, and the other is waiting microk8s itself :/
as for the timeouts, you are absolutely right. removing them :)

    >microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=hostpath-provisioner --timeout=300s

>     >microk8s.kubectl wait -n ingress --for=condition=Ready pod -l name=nginx-ingress-microk8s --timeout=300s

Reviewer: this command should be:
```
microk8s.kubectl wait -n default --for=condition=Ready pod -l name=nginx-ingress-microk8s
```

> 4. Export your microk8s cluster configuration to the local kubeconfig file

Author: not exactly required, you are right. I used it so that i could use k9s for instance

> 4. Export your microk8s cluster configuration to the local kubeconfig file

Author: removing

    >microk8s.config > ~/.kube/config

After runing these commands, we have ensured that dns, http, and nginx ingress are enabled and working properly inside the cluster.

## 5. Installing Codacy

### 1. Helm and Tiller

First we must install `helm` onto the cluster. With `helm`, we can easily install charts and manage the lifecycle of the installed artifacts. One example is the ability to rollback a failed install.

> 1. Assuming you would want to install `helm` version 2.16.3, export the following environment variable

Reviewer 3: Also, this is already described here:

https://github.com/codacy/chart/blob/master/docs/index.md#preparing-to-install-codacy

    >export HELM_VERSION=2.16.3
   1. Please note that we currently do *not* support helm v3.
2. Run this script that will retrieve the desired `helm` version and install it onto the machine's `/usr/local/bin`:
   1. Define a HELM_PKG variable
      >HELM_PKG=helm-v${HELM_VERSION}-linux-amd64.tar.gz
   2. Get the package
      >wget https://get.helm.sh/$HELM_PKG
   3. Extract the tarball
      >tar xvzf $HELM_PKG
   4. Move the binaries to `/usr/local/bin`
      >mv linux-amd64/tiller linux-amd64/helm /usr/local/bin
   5. Delete the extracted tarball directory
      >rm -rvf $HELM_PKG linux-amd64/

3. We must now deploy `tiller` onto the cluster.
   1. First we must set a `serviceaccount` and a `clusterrolebinding` for `tiller` on the microk8s cluster:
      >microk8s.kubectl create serviceaccount --namespace kube-system tiller

      >microk8s.kubectl create clusterrolebinding tiller-cluster-rule --clusterrole=cluster-admin --serviceaccount=kube-system:tiller
   1. Then we can install `tiller`, which will be installed with the same `HELM_VERSION` that you have specified earlier:

>       >helm init --service-account tiller

Author: We only support up to helm 2.16.3 (which is the latest helm2)
I think i need to mention that helm version here then.

### 2. Codacy chart

Please refer to [our chart installation documentation](https://github.com/codacy/chart/blob/master/docs/install.md).

Additionally, any `kubectl` command from [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md) must be executed as a `microk8s.kubectl` command.

#### 1. Default Values

We provide a base (values file)[##microk8s-values.yaml] that you can use for your microk8s Codacy installation. This file provisions a smaller number of component replicas with limited resources (i.e. CPU and memory). Should you want to use this as a template, please pass your `microk8s-values.yaml` file on (4.) of [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md). Lastly, it is important to correctly set your `codacy-api.ingress.hosts.host` if you are using this file as a template.

#### 2. External Databases

If you have decided to go with the [default values for the installation](####1.-Default-Values), your microk8s cluster will contain a PostgreSQL database within itself. This is not recommended as the data stored in the database will not be as durable if you store it in an external database.

> To setup an external database, you should follow the [steps described here](../requirements.md).

Author: I will leave that call for @pauloribeiro-codacy to make :) Just because we don't want to repeat ourselves across the docs so maybe we need to add that info elsewhere (not on microk8s docs)

> To setup an external database, you should follow the [steps described here](../requirements.md).

Reviewer 3: The initial version of the EKS documentation also mentioned a "built-in" database and then making the installation production-ready with an external one. But due to feedback we decided to simplify and include explicit instructions only on how to set up Codacy with the (recommended) external database. So, I suggest that we focus on the recommended path of setting up the connection to the external database that is already part of the pre-requirements. You can check with @paulopontesm regarding what was done for EKS.

In addition to those, you must not forget to include the appropriate configuration blocks in the values file you provide to `helm` during installation.

An example of a configuration block is:

```yaml
global:
  analysisdb:
    create: false
    postgresqlUsername: <--- codacy-db-username --->
    postgresqlDatabase: analysis # You need to create the DB manually
    postgresqlPassword: <--- codacy-db-password --->
    host: <--- codacy-db-host --->
    service:
      port: 5432
```

## 6. Accessing the Codacy UI

Please follow [the Post-install configuration steps described here](../install.md) so that you can access your Codacy installation through the UI.

The crucial configuration step is that you must have the codacy-api ingress and backend urls(`codacy-api.ingress.hosts.host`, `global.codacy.url`, and `global.codacy.backendUrl`) properly configured.

## microk8s-values.yaml

```yaml
global:
  imagePullSecrets:
    - name: docker-credentials
  codacy:

>     url: "${CODACY_PROTOCOL}://${CODACY_URL}"

Author: Good point 👍 you should write them to the file. Fixing this.

      memory: 1Gi
    requests:
      cpu: 100m
      memory: 500Mi

listener:
  replicaCount: 1
  resources:
    limits:
      cpu: 2
      memory: 8Gi
    requests:
      cpu: 1
      memory: 6Gi
  listenerdb:
    persistence:
      enabled: true
      size: 50Gi

  persistence:
    claim:
      size: 50Gi

  nfsserverprovisioner:
    enabled: true
    persistence:
      enabled: true
      size: 100Gi

core:
  replicaCount: 1
  resources:
    limits:
      cpu: 500m
      memory: 1500Mi
    requests:
      cpu: 100m
      memory: 750Mi

engine:
  replicaCount: 1
  resources:
    limits:
      cpu: 2000m
      memory: 15000Mi
    requests:
      cpu: 1000m
      memory: 10000Mi

worker-manager:
  replicaCount: 1
  resources:
    limits:
      cpu: 500m
      memory: 1000Mi
    requests:
      cpu: 100m
      memory: 500Mi
  config:
    workers:
      genericMax: 2
      dedicatedMax: 2
    workerResources:
      limits:
        cpu: 1
        memory: "2Gi"
    pluginResources:
      requests:
        cpu: 0.5
        memory: 1000000000 # 1000Mb

crow:
  replicaCount: 1
  resources:
    limits:
      cpu: 1
      memory: 2Gi
    requests:
      cpu: 0.5
      memory: 1Gi

> postgres:

Reviewer: cc. @h314to is this an option we want to support?

# Setting up a microk8s cluster

Follow the steps below to set up a microk8s cluster from scratch, including all the necessary dependencies and configuration.

## 1. Concepts

### Kubeconfig

> Any tool that operates on top of a cluster needs to know which cluster it is going to work on. This execution context is typically called the `kubeconfig` which is stored in the `~/.kube/config` file. This filed basically defines a list of clusters and it holds several pieces of information per cluster entry: the address of the cluster (e.g. https://104.248.34.242:164309), its user friendly name (e.g. microk8s-cluster). It finally holds a reference to the current context. This is the active cluster against which any kubernetes control (`kubectl`) command you run will get executed.

Reviewer 3: ```suggestion
Any tool that operates on top of a cluster needs to know which cluster it is going to work on. This execution context is typically called the `kubeconfig` which is stored in the `~/.kube/config` file. This file basically defines a list of clusters and it holds several pieces of information per cluster entry:

* The address of the cluster (e.g. https://104.248.34.242:164309).
* The user friendly name of the cluster (e.g. microk8s-cluster).
* A reference to the current context. This is the active cluster against which any kubernetes control (`kubectl`) command you run will get executed.
```

Reviewer 3: Fix typo "filed" and use bullet list for readability.

### Helm and tiller

> Two executables will get installed onto the cluster as part of this process: `helm` and `tiller`. The former is responsible for resolving the configuration of the chart to be installed, while issuing the correct install commands onto the cluster. The latter is responsible for receiving the install commands issued by `helm`, as well as managing the lifecycle of the components that have been installed. `helm` is the client facing side, while `tiller` is the server/cluster facing side.

Reviewer 3: ```suggestion
Two executables will get installed onto the cluster as part of this process: `helm` and `tiller`.

* `helm` is responsible for resolving the configuration of the chart to be installed, while issuing the correct install commands onto the cluster.
* `tiller` is responsible for receiving the install commands issued by `helm`, as well as managing the lifecycle of the components that have been installed.

`helm` is the client facing side, while `tiller` is the server/cluster facing side.
```

Reviewer 3: Use bullet list for readability.

## 2. Prepare your environment

Prepare your environment to set up the microk8s cluster:

1. a machine running Ubuntu 18.04 LTS version.

> 2. establish a remote session (ssh) onto this machine.

Reviewer 3: ```suggestion
1. A machine running Ubuntu 18.04 LTS.
2. Establish a remote SSH session onto this machine.
```

Reviewer 3: Minor style suggestions.

> Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, please jump on to the [Configuring microk8s](###Configuring-microk8s) section.

Reviewer 3: ```suggestion
Assuming that you are starting from a blank slate, the first step is to install microk8s. Otherwise, please jump on to the section [Configuring microk8s](###4.-Configuring-microk8s).
```

Reviewer 3: Make sure the internal links are working on the final version.

## 3. Installing microk8s

1. Make sure the machine is up to date and also has the `nfs-common` package installed

>    >apt update && apt install nfs-common -y

Reviewer 3: ```suggestion

   ```bash
   sudo apt update && apt install nfs-common -y
   ```
```

Reviewer 3: Consider using fenced code blocks throughout the document, specifying the language for code highlighting.

Reviewer 3: The user must run the command as super user.

2. Install microk8s

>    >sudo snap install microk8s --classic --channel=1.15/stable

Reviewer 3: On my Linux machine based on Ubuntu 18.04 LTS, snapd is not installed by default. I had to run:

```bash
sudo apt install snapd
```

>    1. We recommend following the official documentation on how to [install microk8s](https://microk8s.io/docs/).

Reviewer 3: It would be great if we could clarify the exact steps from the microk8s documentation that must be followed. All of them? Only up to a certain point?

## 4. Configuring microk8s

1. First, we must enable the following plugins on microk8s:
    >sudo microk8s.enable dns && sudo microk8s.status --wait-ready

    >sudo microk8s.enable storage && sudo microk8s.status --wait-ready

    >sudo microk8s.enable ingress &&  sudo microk8s.status --wait-ready

2. Now that these plugins have been enabled, we should restart microk8s to ensure a proper bootstrap of the cluster
    >microk8s.stop && microk8s.start && microk8s.status --wait-ready

    Further information on this subject can be [found here](https://github.com/ubuntu/microk8s/issues/493#issuecomment-498167435).

> 3. The plugins are now enabled and the cluster bootstrapped. However, we must still wait for some microk8s internals (dns, http, and ingress) plugins to be ready. Failing to do so can result in pods entering a `CrashLoopBackoff` state:

Reviewer 3: ```suggestion

3. The plugins are now enabled and the cluster bootstrapped. However, we must still wait for some microk8s internals (dns, http, and ingress) plugins to be ready. Failing to do so can result in pods entering a `CrashLoopBackoff` state:
```

Reviewer 3: Empty line between list items.

    >microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=kube-dns --timeout=300s

    >microk8s.kubectl wait -n kube-system --for=condition=Ready pod -l k8s-app=hostpath-provisioner --timeout=300s

    >microk8s.kubectl wait -n ingress --for=condition=Ready pod -l name=nginx-ingress-microk8s --timeout=300s

4. Export your microk8s cluster configuration to the local kubeconfig file
    >microk8s.config > ~/.kube/config

After runing these commands, we have ensured that dns, http, and nginx ingress are enabled and working properly inside the cluster.

## 5. Installing Codacy

> ### 1. Helm and Tiller

Reviewer 3: I would move this section to **Configuring microk8s**, as we're still taking care of the prerequisites and not installing Codacy yet.

Author: But helm is not microk8s specific. It is required to install the codacy chart :/

Reviewer 2: I agree with @pauloribeiro-codacy . This looks a lot like setting up prerequisites to then install codacy. Microk8s is a prerequisite and Helm is another.

First we must install `helm` onto the cluster. With `helm`, we can easily install charts and manage the lifecycle of the installed artifacts. One example is the ability to rollback a failed install.

> 1. Assuming you would want to install `helm` version 2.16.3, export the following environment variable

Reviewer 3: Please see the `helm` version that we are requiring in the `index.md` file (section **Preparing to install Codacy**). We should be consistent here.

    >export HELM_VERSION=2.16.3
   1. Please note that we currently do *not* support helm v3.

> 2. Run this script that will retrieve the desired `helm` version and install it onto the machine's `/usr/local/bin`:

Reviewer 3: Notice that in the `index.md` we are already asking the user to install the required tools by following their documentation. So this step is probably not necessary?

   1. Define a HELM_PKG variable
      >HELM_PKG=helm-v${HELM_VERSION}-linux-amd64.tar.gz
   2. Get the package
      >wget https://get.helm.sh/$HELM_PKG
   3. Extract the tarball
      >tar xvzf $HELM_PKG
   4. Move the binaries to `/usr/local/bin`
      >mv linux-amd64/tiller linux-amd64/helm /usr/local/bin
   5. Delete the extracted tarball directory

>       >rm -rvf $HELM_PKG linux-amd64/

Reviewer 3: Include all the commands in a fenced code block so they're easier to copy+paste at once. We can still explain the main steps of the script below the code block.

3. We must now deploy `tiller` onto the cluster.
   1. First we must set a `serviceaccount` and a `clusterrolebinding` for `tiller` on the microk8s cluster:
      >microk8s.kubectl create serviceaccount --namespace kube-system tiller

      >microk8s.kubectl create clusterrolebinding tiller-cluster-rule --clusterrole=cluster-admin --serviceaccount=kube-system:tiller
   1. Then we can install `tiller`, which will be installed with the same `HELM_VERSION` that you have specified earlier:
      >helm init --service-account tiller
   1. We wait for tiller to be ready

>       >microk8s.kubectl -n kube-system wait --for=condition=Ready pod -l name=tiller --timeout=300s

Reviewer 3: Same recommendation as above, condense all the commands in a single fenced code block and explain what this does below the code block.

### 2. Codacy chart

Please refer to [our chart installation documentation](https://github.com/codacy/chart/blob/master/docs/install.md).

Additionally, any `kubectl` command from [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md) must be executed as a `microk8s.kubectl` command.

#### 1. Default Values

> We provide a base (values file)[##microk8s-values.yaml] that you can use for your microk8s Codacy installation. This file provisions a smaller number of component replicas with limited resources (i.e. CPU and memory). Should you want to use this as a template, please pass your `microk8s-values.yaml` file on (4.) of [our chart installation](https://github.com/codacy/chart/blob/master/docs/install.md). Lastly, it is important to correctly set your `codacy-api.ingress.hosts.host` if you are using this file as a template.

Reviewer 3: Consider providing the `.yaml` file as a standalone file in the GitHub repository just like we did for AKS and EKS:

https://github.com/codacy/chart/tree/master/docs/infrastructure/AKS
https://github.com/codacy/chart/tree/master/docs/infrastructure/EKS

Reviewer 3: Sorry, I confused the files. The ones in the link above are the Terraform backend files and have nothing to do with the `microk8s-values.yaml` file mentioned here.

The file I wanted to mention is actually this one:
https://github.com/codacy/chart/blob/master/codacy/values-production.yaml

> #### 2. External Databases

Reviewer 3: Here we already have instructions on how to create the external database:

https://github.com/codacy/chart/blob/master/docs/requirements.md#postgresql-server-setup

Are there steps that are significantly different when using microk8s? If not, we can assume that the user already has an external database as per the requirements.

Author: The only thing missing is to mention in the docs to add the specific values to the values file i guess. To have something like an example block. Let me iterate on this. :)

If you have decided to go with the [default values for the installation](####1.Default-Values), your microk8s cluster will contain a postgresql database within itself. This is not recommended as the data stored in the database will not be as durable if you store it in an external database.

> To use an external database, you should follow the following steps:

Reviewer 3: ```suggestion
If you have decided to go with the [default values for the installation](####1.-Default-Values), your microk8s cluster will contain a PostgreSQL database within itself. This is not recommended as the data stored in the database will not be as durable if you store it in an external database.

Follow the steps below to use an external database:
```

Reviewer 3: Minor style adjustments.

4. The database cluster must have the following databases
   1. activities
   2. hotspots
   3. listener
5. In the values file that you provide during the `helm` installation, define the following for the databases described above:
   1. `<database>db.create=false` - disable the creation of this database inside the cluster.
   2. `<database>db.postgresqlUsername` - the username that will have access to this database.
   3. `<database>db.postgresqlPassword` - the passworf for the given user.
   4. `<database>db.host` - hostname of the external database.
6. Note that the provided user must be the `owner` of the database(s):
    >ALTER ROLE \<user> WITH CREATEDB;

    Then you can connect as that user and create the databases:
    ```sql
    CREATE DATABASE accounts WITH OWNER=<user>;
    CREATE DATABASE analysis WITH OWNER=<user>;
    CREATE DATABASE results WITH OWNER=<user>;
    CREATE DATABASE metrics WITH OWNER=<user>;
    CREATE DATABASE filestore WITH OWNER=<user>;
    CREATE DATABASE jobs WITH OWNER=<user>;
    CREATE DATABASE activities WITH OWNER=<user>;
    CREATE DATABASE hotspots WITH OWNER=<user>;
    CREATE DATABASE listener WITH OWNER=<user>;
    ```

> ## 6. Accessing the Codacy UI

Reviewer 3: This should be tied better with this section:

https://github.com/codacy/chart/blob/master/docs/index.md#post-install-configuration

In the previous section, you should have defined values for `codacy-api.ingress.hosts.host`, `global.codacy.url`, and `global.codacy.backendUrl`. This value will be the address where you can access the Codacy UI from your web browser.

## 7. Troubleshooting

> ### 1. AWS EC2 specific issues

Reviewer 3: I think this may be quite specific to the test that you did and may confuse customers since we also provide instructions to install Codacy on Amazon EKS.

I was thinking that customers using microk8s will use any kind of infrastructure, most probably one that is managed by them?

Consider removing this section.

Author: 👍 removing

1. We had to configure the instance's list of hosts with localhost. This was because our machine was in fact an AWS EC2 instance, and therefore we must allow the correct resolution of the local microk8s cluster on the kubeconfig. We did so by running

    >sudo sed -i -e "1s/127.0.0.1 localhost.*/127.0.0.1 localhost $(hostname)/" /etc/hosts

2. Traffic forwarding must be configured on the machine so that the local microk8s cluster pods can reach the internet, again because of using an AWS EC2 instance. This can be done by running
    >sudo iptables -P FORWARD ACCEPT

    (more info on this subject can be [found here](https://microk8s.io/docs/troubleshooting#common-issues)).

> ## microk8s-values.yaml

Reviewer 3: See comment above about providing this as a standalone file. :)

Author: I am not 100% sure what to do about it :/ suggestions?

Reviewer 3: I confused the files in my comment above, sorry.

What I meant is that we are providing the following file in the GitHub repository itself:

https://github.com/codacy/chart/blob/master/codacy/values-production.yaml

We could also provide this `microk8s-values.yaml` in the same way so that it's a bit easier for users to download and use it - and it also makes this documentation shorter.

I suggest calling the file `values-production-microk8s.yaml` so it is ordered alphabetically and stays close to the `values-production.yaml` one.

Author: Good idea. I wouldn't call it 'production' however. It is mostly just a placeholder/template/starting point. But i have moved it to live alongside the other values files :)
